leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/impl/CaliforniumEndpointsManager.java

```diff
@@ -0,0 +1,198 @@
+import org.eclipse.californium.core.network.Endpoint;
+import org.eclipse.californium.core.network.config.NetworkConfig;
+import org.eclipse.californium.scandium.DTLSConnector;
+import org.eclipse.californium.scandium.config.DtlsConnectorConfig;
+import org.eclipse.californium.scandium.config.DtlsConnectorConfig.Builder;
+import org.eclipse.californium.scandium.dtls.pskstore.StaticPskStore;
+import org.eclipse.leshan.SecurityMode;
+import org.eclipse.leshan.client.servers.EndpointsManager;
+import org.eclipse.leshan.client.servers.Server;
+import org.eclipse.leshan.client.servers.ServerInfo;
+import org.eclipse.leshan.core.californium.EndpointFactory;
+import org.eclipse.leshan.core.request.Identity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CaliforniumEndpointsManager implements EndpointsManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CaliforniumEndpointsManager.class);
+
+    private boolean started = false;
+
+    private Endpoint currentEndpoint;
+    private Builder dtlsConfigbuilder;
+    private NetworkConfig coapConfig;
+    private InetSocketAddress localAddress;
+    private CoapServer coapServer;
+    private EndpointFactory endpointFactory;
+
+    public CaliforniumEndpointsManager(CoapServer coapServer, InetSocketAddress localAddress, NetworkConfig coapConfig,
+            Builder dtlsConfigBuilder, EndpointFactory endpointFactory) {
+        this.coapServer = coapServer;
+        this.localAddress = localAddress;
+        this.coapConfig = coapConfig;
+        this.dtlsConfigbuilder = dtlsConfigBuilder;
+        this.endpointFactory = endpointFactory;
+    }
+
+    @Override
+    public synchronized Identity createEndpoint(ServerInfo serverInfo) {
+        // Clear previous endpoint
+        if (currentEndpoint != null) {
+            coapServer.getEndpoints().remove(currentEndpoint);
+            currentEndpoint.destroy();
+        }
+
+        // Create new endpoint
+        Identity server;
+        if (serverInfo.isSecure()) {
+            Builder newBuilder = cloneDtlsConfigBuilder(dtlsConfigbuilder);
+
+            // Support PSK
+            if (serverInfo.secureMode == SecurityMode.PSK) {
+                StaticPskStore staticPskStore = new StaticPskStore(serverInfo.pskId, serverInfo.pskKey);
+                newBuilder.setPskStore(staticPskStore);
+            }
+            // TODO add support for RPK and X509
+
+            if (endpointFactory != null) {
+                currentEndpoint = endpointFactory.createSecuredEndpoint(newBuilder.build(), coapConfig, null);
+            } else {
+                currentEndpoint = new CoapEndpoint(new DTLSConnector(newBuilder.build()), coapConfig, null, null);
+            }
+            server = Identity.psk(serverInfo.getAddress(), serverInfo.pskId);
+        } else {
+            if (endpointFactory != null) {
+                currentEndpoint = endpointFactory.createUnsecuredEndpoint(localAddress, coapConfig, null);
```

Comment:
what about passing the ObservationStore to both endpoints?

Reply:
The `ObservationStore` is only required for the CoAP-client of such an observation. Though LWM2M TS doesn't define, that a LWM2M-device-client is allowed to establish an observation of a LWM2M-server resource, I think, the `ObservationStore` is not used at the LWM2M-device-client.

Reply:
I agree with @boaks, I think `ObservationStore` is not needed at LWM2M-client side.

Reply:
oh, I got confused